File: product-brain/src/product_brain/repair.py
from __future__ import annotations

import subprocess
from pathlib import Path

from .config import Config, load as load_config
from .index import list_records, read_records, write_record
from .blocks.edge_mine import validate_citations


def _head_exists(repo: Path, path: str) -> bool:
    return (repo / path).exists()


def repair_repo(config: Config, repo_name: str) -> dict:
    repo_cfg = config.repo(repo_name)
    repo_path = repo_cfg.path
    records = read_records(repo_path, repo_name)
    bullets_dropped = 0
    files_marked_deleted = 0
    stale_tagged = 0

    pr_numbers_seen: set[int] = set()
    for r in records.values():
        pr_numbers_seen.update(r.prs)

    for ticket_id, rec in records.items():
        edges, dropped_e = validate_citations(rec.edge_cases_handled, repo_path, pr_numbers_seen)
        gaps, dropped_g = validate_citations(rec.known_gaps, repo_path, pr_numbers_seen)
        bullets_dropped += dropped_e + dropped_g

        for f in rec.files:
            if f.change != "deleted" and not _head_exists(repo_path, f.path):
                f.change = "deleted"
                files_marked_deleted += 1

        new_gaps = []
        for g in gaps:
            if "TODO" in g.source and ":" in g.source:
                path = g.source.split(":")[0]
                if not _head_exists(repo_path, path):
                    g.text = f"[stale] {g.text}"
                    stale_tagged += 1
            new_gaps.append(g)

        rec.edge_cases_handled = edges
        rec.known_gaps = new_gaps
        write_record(repo_path, rec)

    return {
        "repo": repo_name,
        "tickets": len(records),
        "bullets_dropped": bullets_dropped,
        "files_marked_deleted": files_marked_deleted,
        "stale_gaps_tagged": stale_tagged,
    }


def main():
    config = load_config()
    for r in config.repos:
        print(repair_repo(config, r.name))


if __name__ == "__main__":
    main()

File: product-brain-ts/src/repair.ts
import { existsSync } from "node:fs";
import path from "node:path";
import { getTestAdapter } from "./adapters/index.js";
import { detectGaps } from "./blocks/coverage-gap.js";
import { stabilitySignals, validateCitations } from "./blocks/edge-mine.js";
import type { Config } from "./config.js";
import type { RunResult } from "./models.js";
import { readRecords } from "./records/read.js";
import { writeRecord } from "./records/write.js";

export interface RepairResult {
  repo: string;
  tickets: number;
  bullets_dropped: number;
  files_marked_deleted: number;
  stale_gaps_tagged: number;
  stability_refreshed: number;
}

function headExists(repo: string, p: string): boolean {
  return existsSync(path.join(repo, p));
}

export async function repairRepo(config: Config, repoName: string): Promise<RepairResult> {
  const repoCfg = config.repo(repoName);
  const sourcePath = repoCfg.path;
  const brainRoot = config.brainRoot;
  const records = readRecords(brainRoot, repoName);
  const testAdapter = getTestAdapter(config.testAdapter, config);

  let bulletsDropped = 0;
  let filesMarkedDeleted = 0;
  let staleTagged = 0;
  let stabilityRefreshed = 0;

  const prNumbersSeen = new Set<number>();
  const caseIdsSeen = new Set<string>();
  for (const r of Object.values(records)) {
    for (const n of r.prs) prNumbersSeen.add(n);
    for (const c of r.testCases) caseIdsSeen.add(c.id);
  }

  for (const rec of Object.values(records)) {
    const e = validateCitations(rec.edgeCasesHandled, sourcePath, prNumbersSeen, caseIdsSeen);
    const g = validateCitations(rec.knownGaps, sourcePath, prNumbersSeen, caseIdsSeen);
    const q = validateCitations(rec.qaEdges, sourcePath, prNumbersSeen, caseIdsSeen);
    bulletsDropped += e.dropped + g.dropped + q.dropped;

    if (testAdapter && rec.testCases.length) {
      try {
        const freshRuns: RunResult[] = [];
        for (const c of rec.testCases) {
          const runs = await testAdapter.fetchRunHistory(c.id);
          freshRuns.push(...runs);
          c.recentFailures = runs.filter((r) => r.status === "failed" || r.status === "blocked").length;
        }
        rec.stabilitySignals = stabilitySignals(rec.testCases, freshRuns);
        stabilityRefreshed += 1;
        if (rec.edgeCasesHandled.length) {
          rec.coverageGaps = await detectGaps(rec.edgeCasesHandled, rec.testCases);
        }
      } catch {
        // adapter unreachable; preserve existing data
      }
    }

    for (const f of rec.files) {
      if (f.change !== "deleted" && !headExists(sourcePath, f.path)) {
        f.change = "deleted";
        filesMarkedDeleted += 1;
      }
    }

    const newGaps = g.kept.map((b) => {
      if (b.source.includes("TODO") && b.source.includes(":")) {
        const filePath = b.source.split(":")[0] ?? "";
        if (!headExists(sourcePath, filePath)) {
          staleTagged += 1;
          return { ...b, text: `[stale] ${b.text}` };
        }
      }
      return b;
    });

    rec.edgeCasesHandled = e.kept;
    rec.knownGaps = newGaps;
    rec.qaEdges = q.kept;
    writeRecord(brainRoot, rec);
  }

  return {
    repo: repoName,
    tickets: Object.keys(records).length,
    bullets_dropped: bulletsDropped,
    files_marked_deleted: filesMarkedDeleted,
    stale_gaps_tagged: staleTagged,
    stability_refreshed: stabilityRefreshed,
  };
}
